### PR TITLE
feat: Collect logs from all pods

### PR DESCRIPTION
**Pull request by rishabh96b**
This patch adds a collector named `AllLogs` whose purpose is to collect logs from all the pods in given list of namespaces(or if not specified, from all namespaces).
In the collector specification, this will look like - 
```
collectors:
  allLogs:
    collectorName: allPodLogs
    namespaces:
      - default
      - dev
      - prod
    selector:
      - env=dev
      - type=test
```
Above specification will make troubleshoot collect pod logs from `default`, `dev`, `prod` namespaces with the selectors `env=dev` & `type=test`

Example of the data produced by running the collector for all `ns`:

```
allPodLogs
├── default
│   ├── nginx-deploy-8588f9dfb-5gz8w.log
│   └── nginx-deploy-8588f9dfb-f8f2d.log
├── dev
│   └── elastic-dev-es-default-0
│       ├── elastic-internal-init-filesystem.log
│       └── elasticsearch.log
├── elastic-system
│   ├── elastic-operator-0.log
│   └── elastic-operator-0-previous.log
├── kube-system
│   ├── coredns-558bd4d5db-4c42v.log
│   ├── coredns-558bd4d5db-mstqq.log
│   ├── etcd-kind-control-plane.log
│   ├── kindnet-vsnpv.log
│   ├── kube-apiserver-kind-control-plane.log
│   ├── kube-controller-manager-kind-control-plane.log
│   ├── kube-proxy-zmnf9.log
│   └── kube-scheduler-kind-control-plane.log
└── local-path-storage
 ```
For more info, check [README.d2iq.md](https://github.com/mesosphere/troubleshoot/blob/rb/all-pods-logs/README.d2iq.md#alllogs-collector)

### CODE REVIEW

File: pkg/collect/all_logs.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		logsCollector.Namespaces = getNamespaceNames(namespaceList)
+	}
+
+	for _, namespace := range logsCollector.Namespaces {
+		pods, podsErrors := listPodsInNamespace(ctx, client, namespace, logsCollector.Selector)
+		if len(podsErrors) > 0 {
+			output.SaveResult(c.BundlePath, getAllLogsErrorsFileName(logsCollector), marshalErrors(podsErrors))
+		}
+
+		if len(pods) > 0 {
```
Comment (mhrabovcin) [inline pkg/collect/all_logs.go:38]: This `if` is unnecessary, it is enough to do `for` loop. If there are no items nothing will get executed.

**Comment by rishabh96b** [comment]
Yes, for sure. For the time being, I've updated the description for a quick look as well.

File: pkg/collect/all_logs.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+							}
+							output.SaveResult(c.BundlePath, key, marshalErrors([]string{err.Error()}))
+							if err != nil {
+								return nil, err
+							}
+							continue
+						}
+						for k, v := range podLogs {
+							output[k] = v
+						}
```
Comment (mhrabovcin) [inline pkg/collect/all_logs.go:68]: This code block is almost identical to the one bellow. Can we change the code so we're not duplicating it?

File: pkg/collect/util.go [hunk 3]
```diff
@@ -43,6 +43,10 @@ func DeterministicIDForCollector(collector *troubleshootv1beta2.Collect) string
 		unsafeID = fmt.Sprintf("logs-%s-%s", collector.Logs.Namespace, selectorToString(collector.Logs.Selector))
 	}
 
+	if collector.AllLogs != nil {
+		unsafeID = fmt.Sprintf("logs-%s-%s", namespaceToString(collector.AllLogs.Namespaces), selectorToString(collector.Logs.Selector))
```
Comment (mhrabovcin) [inline pkg/collect/util.go:47]: What if `Namespaces` and `Selector` are empty?

File: pkg/collect/all_logs.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	ctx := context.Background()
+
+	// Set default logCollector Name if empty
+	logCollectorName := ""
+
+	if logsCollector.Name == "" {
+		logCollectorName = "allPodLogs"
+	} else {
+		logCollectorName = logsCollector.Name
+	}
```
Comment (mhrabovcin) [inline pkg/collect/all_logs.go:33]: logCollectorName := "allPodLogs"
if logsCollector.Name != "" {
	logCollectorName = logsCollector.Name
}

File: pkg/collect/all_logs.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+						if err != nil {
+							return nil, err
+						}
+						continue
+					}
+					for k, v := range containerLogs {
+						output[k] = v
+					}
+				}
+			}
```
Comment (mhrabovcin) [inline pkg/collect/all_logs.go:96]: This code block is very similar to the one above. Is there any chance we can dedupe these?

Comment (rishabh96b) [reply pkg/collect/all_logs.go:96]: Tried de-duping the looping block but the common func seems to have more than 10 arguments to make it work which does not look good to me. It looks clear this way that we are either looping through all the containers within a pod or specified containers.

File: pkg/collect/all_logs.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	return namespaceNames
+}
+
+func getAllLogsErrorsFileName(logsCollector *troubleshootv1beta2.AllLogs) string {
+	if len(logsCollector.Name) > 0 {
+		return fmt.Sprintf("%s/errors.json", logsCollector.Name)
+	} else if len(logsCollector.CollectorName) > 0 {
+		return fmt.Sprintf("%s/errors.json", logsCollector.CollectorName)
+	}
+	// TODO: random part
```
[the remark this answers is not in the document]
Comment (rishabh96b) [reply pkg/collect/all_logs.go:143]: Removed already!

**Review by mhrabovcin — APPROVED** [review]
Very nice contribution!

**Pull request merged**